chirp/drivers/tdh8.py

```diff
@@ -0,0 +1,2479 @@
+}
+
+
+def _ident_radio(radio):
+    for magic in radio._idents:
+        error = None
+        try:
+            data = _do_ident(radio, magic)
+            return data
+        except errors.RadioError as e:
+            # LOG.error("tdh8._ident_radio: %s", e)
+            error = e
+            time.sleep(2)
+
+    if error:
+        raise error
+    raise errors.RadioError("Radio did not respond")
+
+
+def _do_download(radio):
+    data = _ident_radio(radio)
+    # radio_version = _get_radio_firmware_version(radio)
+    # LOG.info("Radio Version is %s" % repr(radio_version))
+    append_model = False
+    # HAM OR GMRS
+    LOG.info("Radio mode is " + MODE_RADIO)
+    LOG.info("Chirp choose mode is " + radio.ident_Mode)
+    # The Ham and Gmrs modes are subclasses of this model tdh8.
+    # When the Ham or Gmrs mode is selected,
+    # a double verification operation will be performed to
+    # obtain the current mode code of the walkie-talkie and
+    # the mode code currently selected by the chirp for comparison.
+    # If the two are consistent, enter This mode,
+    # otherwise an inconsistent error will be prompted.
+    if MODE_RADIO == radio.ident_Mode:
+        LOG.info("Successful match.")
+    else:
+        msg = (
+            "Please check whether the walkie-talkie mode is consistent with "
+            "the selected model mode.")
+        raise InvalidValueError(msg)
+
+    # Main block
+    LOG.info("Downloading...")
+
+    for i in range(0, radio._memsize, 0x20):
+        block = _read_block(radio, i, 0x20)
+        data += block
+        _do_status(radio, i)
+    _do_status(radio, radio._memsize)
+    LOG.info("done.")
+
+    if append_model:
+        data += radio.MODEL.ljust(8)
+
+    # LOG.debug("done.")
+    return memmap.MemoryMapBytes(data)
+
+
+def _exit_write_block(radio):
+    serial = radio.pipe
+    try:
+        serial.write(b"E")
+
+    except:
+        raise errors.RadioError("Radio refused to exit programming mode")
+
+
+def _write_block(radio, addr, data):
+    serial = radio.pipe
+    cmd = struct.pack(">cHb", b'W', addr, 0x20)
+    data = radio.get_mmap()[addr + 8: addr + 40]
+    cmd_str = hex(sum(data) & 0xff)
+    data_bused = chr(int(cmd_str, 16))
+
+    # This is because when our cmd_str is converted to decimal,
+    # sometimes num will be greater than 128, but chr(num).encode('utf-8') gets
+    # two-byte ASCII characters, each in ASCII encoding The character
+    # encoding is one byte, and only 0-127, and the encoding of the 128-255
+    # part belongs to Extended ASCII, and the characters in the extended
+    # ascii part greater than 128 are encoded by two bytes, and two
+    # characters will be obtained Bytes, what is needed is an
+    # encoding that supports Extended ASCII characters and uses 1 Byte
+    # fixed-size encoding, such as ISO 8859-1, also known as latin1:
+    # >>> chr(165)
+    # ¥
+    # >>> chr(165).encode('latin1')
+
+    data += data_bused.encode('latin1')
```

Comment:
You should pretty much never be doing this kind of encode operation.

Reply:
If I use encode() directly, it will report an error.

Reply:
Right, because it's not valid `UTF-8`. It's not valid `latin1` either, it's just that `latin1` is a direct 1:1 byte-for-byte conversion. You should be using `int` and `bytes` types for binary data, not `str`. I gave you the solution above.